### ISSUE TITLE
missing entrypoint

### ISSUE THREAD

**Issue by smadigan-dexcom** [GoogleCloudPlatform/google-cloud-dotnet#1028]
`The native method "grpcsharp_request_call_context_create" does not exist`

I'm banging my head against the wall on this one.  I've tried upgrading and downgrading in just about every combination of the various dependencies with no luck.  Can someone shed light on when this entry point appeared/disappeared, what likely conflicts might be?  I am using the latest pre-release versions of all components.

   `at Grpc.Core.Internal.UnmanagedLibrary.GetNativeMethodDelegate[T](String methodName)
   at Grpc.Core.Internal.NativeMethods..ctor(UnmanagedLibrary library)
   at Grpc.Core.Internal.NativeExtension..ctor()
   at Grpc.Core.Internal.NativeExtension.Get()
   at Grpc.Core.GrpcEnvironment.GrpcNativeInit()
   at Grpc.Core.GrpcEnvironment..ctor()
   at Grpc.Core.GrpcEnvironment.AddRef()
   at Grpc.Core.Channel..ctor(String target, ChannelCredentials credentials, IEnumerable`1 options)
   at Google.Api.Gax.Grpc.ChannelPool.GetChannel(ServiceEndpoint endpoint, ChannelCredentials credentials)
   at Google.Cloud.PubSub.V1.SubscriberClient.Create(ServiceEndpoint endpoint, SubscriberSettings settings)

**Comment by jskeet** [comment]
Could you tell us more about how you're trying to run your code please? In particular:

- What kind of application is this?
- How are you building and deploying it?
- Which version of .NET are you targeting?
- What processor architecture are you running on?

cc @jtattermusch

**Comment by smadigan-dexcom** [comment]
- This is a windows service, however I'm running in a test environment, and the exception occurs while under test.  I have not deployed the service because it's not running under test.  The test runner is the Jetbrains/resharper NUnit runner.
- Built under Visual Studio
- Targeting .Net 4.5.2 (not core)
- Any CPU build, 64bit processor

Thanks in advance

**Comment by jskeet** [comment]
Hmm... I wonder whether the test runner is the cause here. If it's copied the .NET assemblies but not the native code, that could explain things. Out of interest, does the same thing happen if you use the NUnit console runner from the command line? Are you able to reproduce this with a tiny library that basically does nothing other than create the channel? (If I can reproduce this easily, I can stop asking questions :)

@jtattermusch: Any suggestions of further diagnostics we could perform?

**Comment by smadigan-dexcom** [comment]
I created a simple console app that just sets credentials and creates a pubsub client.  No problems.  I added a test class to run the console app under the test runner.  No problems.  Added the BigQuery libraries (as in the problem app), run again, no problems.  I'll start comparing the differences between the problem app and the no-problem and see where that leads.  Any other advice or diagnostic approaches appreciated.

**Comment by jskeet** [comment]
BigQuery should be completely irrelevant to this, as that's a REST API - I wouldn't expect gRPC to get involved there at all. It's PubSub that's the "problem" API here, as that uses gRPC. So I'd *only* use PubSub in the tests. Is there anything in the broken code that changes current directory or loads anything with reflection? I'd double check that the console app is also AnyCPU - and possibly *try* changing the service to be explicitly x64 rather than AnyCPU. Hmm. We'll get there :)

**Comment by smadigan-dexcom** [comment]
no changes to current directory; there's a call to Assembly.GetExecutingAssembly() to find the directory where the .json credentials file lives, but no loading of assemblies with reflection.  Console app is built/run as AnyCPU.  Changing to x64, no change, still no problem.

**Comment by jskeet** [comment]
Sorry, I meant try changing the broken service to x64.

It's been a while since I've done any service work - is that built as an executable, or just a class library? If it's an executable, what kind of executable?

**Comment by smadigan-dexcom** [comment]
Thanks to your guidance, I think I've proven to myself that this is all an artifact of the test environment.  The original test project had references to the Google libraries due to the kinds of tests in that project.  I created a new test project that only references the class under test and everything is fine, no problems.  That's not to say that there isn't a problem here worth figuring out, but at least I'm not stuck any more.  I'm going to keep the environment (original test project) that causes this and see if I can't isolate the cause.

**Comment by jskeet** [comment]
Great - your efforts in going the extra mile are much appreciated, and I'm glad you're unblocked for now :)

**Comment by jtattermusch** [comment]
@smadigan-dexcom  it would help seeing the structure of your projects.  
In general:
1. try checking that grpc_csharp_ext.X.dll  is copied to the output folder (if not, that's a problem)
2. make sure that the project you're executing has Grpc.Core as its dependency (that's the one that copies the native library to the output dir - it should work transitively, too, but it's worth checking).

**Comment by smadigan-dexcom** [comment]
Not sure how to communicate the structure of the projects -- this is a solution with over a hundred projects (don't ask!)  The test project in this case references both projects that have references to the Google libraries as well as other projects that have references.

1.  the grpc_csharp_ext libraries (x86 and 64) are copied to the output directory
2.  the test project and the project with the class under test have direct references to Grpc.Core

I continue to be mystified.  I've added references to Google libraries to the test project that is succeeding (the new one), and it still works.  Whatever difference exists between the failing and succeeding projects, it's either very subtle or too obvious, either way I haven't figured it out yet.

**Comment by jskeet** [comment]
Have you tried comparing just the output folders of the working and broken ones? (Is there an app.config that could be different?)

**Comment by jskeet** [comment]
Any update to this, or is it worth closing as effectively unreproducible?

**Comment by smadigan-dexcom** [comment]
I think we can close it as such for now.  If I can ever reproduce it I’ll open a new issue.

Thanks!

**Comment by jskeet** [comment]
Righto.

**Comment by kassemshehady** [comment]
Same here http://track.kryptonz.com/

**Comment by jskeet** [comment]
@kassemshehady: Are you able to reproduce this reasonably simply? We'd love to fix it if we can get hold of it properly...

**Comment by jskeet** [comment]
Just as a starting point, could you check which version of Grpc.Core you're using? We're targeting 1.2, but I'm aware that 1.3 is out - which *should* be fine, but I haven't personally used 1.3 much.

**Comment by kassemshehady** [comment]
I am using Grpc.Core 1.3

**Comment by kassemshehady** [comment]
The project is working locally, but on a google VM it gives this error.

**Comment by jskeet** [comment]
Okay. This was a new function in the native library, by the looks of it. Which platform is this failing on? I'm wondering whether the nuget package has an old version of the native library for one platform, or something like that.

cc @jtattermusch who may know more.

**Comment by kassemshehady** [comment]
I deploy another project using the same packages and it is working fine , but in this project the error occurs remotely

**Comment by kassemshehady** [comment]
I replace grpc_csharp_ext.x64.dll from the other project in th bin directory and its working now

**Comment by jskeet** [comment]
Okay, so to clarify, you were using Windows when deploying within Google as well? Did you happen to check the size of the old/new files? (Perhaps upgrading Grpc.Core doesn't copy the new native libraries in some cases...)

**Comment by kassemshehady** [comment]
OK i found it 
The project that fails have in the project file 
<Target Name="EnsureNuGetPackageBuildImports" BeforeTargets="PrepareForBuild">
    <PropertyGroup>
      <ErrorText>This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is {0}.</ErrorText>
    </PropertyGroup>
    <Error Condition="!Exists('..\..\packages\Grpc.Core.1.0.1\build\net45\Grpc.Core.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\..\packages\Grpc.Core.1.0.1\build\net45\Grpc.Core.targets'))" />
    <Error Condition="!Exists('..\..\packages\Microsoft.Net.Compilers.2.1.0\build\Microsoft.Net.Compilers.props')" Text="$([System.String]::Format('$(ErrorText)', '..\..\packages\Microsoft.Net.Compilers.2.1.0\build\Microsoft.Net.Compilers.props'))" />
    <Error Condition="!Exists('..\..\packages\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.1.0.4\build\net45\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.props')" Text="$([System.String]::Format('$(ErrorText)', '..\..\packages\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.1.0.4\build\net45\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.props'))" />
  </Target>
  <Import Project="..\..\packages\Grpc.Core.1.0.1\build\net45\Grpc.Core.targets" Condition="Exists('..\..\packages\Grpc.Core.1.0.1\build\net45\Grpc.Core.targets')" />

This code is copied from the sample code provided by google cloud platform using .net

While the working doesn't have this code about Grpc

**Comment by jskeet** [comment]
Could you point to that sample? We should update it? I wouldn't expect a specific version of Grpc.Core there - you should only need to include the NuGet package, and it will be fine.

cc @SurferJeffAtGoogle who may know more.

**Comment by kassemshehady** [comment]
I use this project Book shelf Pub/Sub as a template for my project from about 5 month https://github.com/GoogleCloudPlatform/getting-started-dotnet/tree/master/aspnet/5-pubsub/bookshelf

**Comment by jskeet** [comment]
Right - it looks like that isn't in the csproj any more...
